viadot/sources/cloud_for_customers.py

```diff
@@ -5,56 +5,169 @@
+
+
+def change_to_meta_url(url):
+    meta_url = ""
+    start = url.split(".svc")[0]
+    ending = url.split("/")[-1]
+    e = ending.split("?")[0]
+    meta_url = start + ".svc/$metadata?entityset=" + e
+    return meta_url
+
+
+def response_to_entity_list(dirty_json, url):
+    metadata_url = change_to_meta_url(url)
+    column_maper_dict = column_mapper(metadata_url)
+    entity_list = []
+    for element in dirty_json["d"]["results"]:
+        new_entity = {}
+        for key, object_of_interest in element.items():
+            if (
+                key != "__metadata"
+                and key != "Photo"
+                and key != ""
+                and key != "Picture"
+            ):
+                # skip the column which contain nested structure
+                if "{" not in str(object_of_interest):
+                    new_key = column_maper_dict.get(key)
+                    if new_key:
+                        new_entity[new_key] = object_of_interest
+                    else:
+                        new_entity[key] = object_of_interest
+        entity_list.append(new_entity)
+    return entity_list
+
+
+class APIError(Exception):
+    pass
 
-    Args:
-        url (str, optional): The url to the API. Defaults to None.
-        endpoint (str, optional): The endpoint of the API. Defaults to None.
-        params (Dict[str, Any]): The query parameters like filter by creation date time. Defaults to json format.
-    """
 
+class CloudForCustomers(Source):
     def __init__(
         self,
         *args,
+        direct_url: str = None,
         url: str = None,
         endpoint: str = None,
         params: Dict[str, Any] = {},
-        **kwargs
+        **kwargs,
     ):
+        """
+        Fetches data from Cloud for Customer.
+
+        Args:
+            direct_url (str, optional): The url to the API in case of prepared report. Defaults to None.
+            url (str, optional): The url to the API. Defaults to None.
+            endpoint (str, optional): The endpoint of the API. Defaults to None.
+            params (Dict[str, Any]): The query parameters like filter by creation date time. Defaults to json format.
+        """
         super().__init__(*args, **kwargs)
         credentials = local_config.get("CLOUD_FOR_CUSTOMERS")
-        self.api_url = url or credentials["server"]
+        self.url = url or credentials["server"]
         self.query_endpoint = endpoint
         self.params = params
         self.params["$format"] = "json"
-        self.auth = (credentials["username"], credentials["password"])
+        access = {"username": None, "password": None}
+        if direct_url:
+            if "my336539" in direct_url:
+                access = credentials.get("QA")
+            elif "my341115" in direct_url:
+                access = credentials.get("Prod")
+        elif url:
+            if "my336539" in url:
+                access = credentials.get("QA")
+            elif "my341115" in url:
```

Comment:
These are names of private C4C instances and should be removed from here.

Also, I don't like the name `access`. We normally refer to credentials as `credentials`. In this case, you could just add a prefix, such as `c4c_credentials = credentials.get("QA")`, etc.

Regarding reading the credentials themselves, this is implemented in the base `Source` class. If the user wants to use different credentials, they should pass a different dictionary containing them to the `credentials` parameter
eg, like this:
```python
qa_credentials = local_config.get("CLOUD_FOR_CUSTOMERS")["QA"]
my_source = CloudForCustomers(credentials=qa_credentials)
```
You should probably set `QA` credentials as the default.